specfile/rpm.py

```diff
@@ -0,0 +1,253 @@
+                        m.group("u") == "=",
+                    )
+                )
+            elif result:
+                if result[-1].body:
+                    result[-1].body += "\n"
+                result[-1].body += line
+        return result
+
+    @classmethod
+    def dump(cls) -> List[Macro]:
+        """
+        Dumps all macros defined in the global context.
+
+        This is not 100% accurate, since macros can be defined multiple times,
+        but only the last definition is listed.
+
+        Returns:
+            List of `Macro` instances.
+        """
+        with capture_stderr() as stderr:
+            rpm.expandMacro("%dump")
+        return cls._parse([line.decode() for line in stderr])
+
+    @staticmethod
+    def expand(expression: str) -> str:
+        """
+        Expands an expression in the global context.
+
+        Args:
+            expression: Expression to expand.
+
+        Returns:
+            Expanded expression.
+
+        Raises:
+            RPMException, if expansion error occurs.
+        """
+        try:
+            with capture_stderr() as stderr:
+                return rpm.expandMacro(expression)
+        except rpm.error as e:
+            raise RPMException(stderr=stderr) from e
+
+    @classmethod
+    def remove(cls, macro: str) -> None:
+        """
+        Removes all definitions of a macro in the global context.
+
+        Args:
+            macro: Macro name.
+        """
+        while True:
+            rpm.delMacro(macro)
+            try:
+                if cls.expand(f"%{macro}") == f"%{macro}":
+                    break
+            except RPMException:
+                # the macro can't be expanded, but it still exists
+                continue
+
+    @classmethod
+    def define(cls, macro: str, body: str) -> None:
+        """
+        Defines a macro in the global context.
+
+        Removes all existing definitions first. It is not possible to define a macro
+        with options. The new macro will always have `MacroLevel.OLDSPEC` level.
+
+        Args:
+            macro: Macro name.
+            body: Macro body.
+        """
+        cls.remove(macro)
+        rpm.addMacro(macro, body)
+
+    @classmethod
+    def reinit(cls, level_threshold: MacroLevel = MacroLevel.RPMRC) -> None:
+        """
+        Reinitializes macros in the global context.
+
+        Args:
+            level_threshold: Only macros up to this level remain defined.
+        """
+        # reset everything, including macros
+        rpm.reloadConfig()
+        for macro in cls.dump():
+            if macro.level > level_threshold:
+                cls.remove(macro.name)
+
+
+class RPM:
+    @staticmethod
+    def parse(
+        s: str, sourcedir: Path, macros: Optional[List[Tuple[str, str]]] = None
+    ) -> rpm.spec:
+        """
+        Parses the content of a spec file.
+
+        Args:
+            s: String representing the content of a spec file.
```

Comment:
Is there a reason behing this naming? I'd probably much rather see `content` rather than `s`.

Reply:
Ok, I'll change it.